File: src/renderer/export/css-generator.js
// ============================================================
// css-generator.js - HTML/CSS 生成の純粋ヘルパー群
// renderer.js(HtmlRenderer) から共有。副作用なし・DOM非依存で単体テストしやすい。
// ============================================================
export const ANIM_CSS = `
        @keyframes fadeIn { from { opacity: 0; } to { opacity: 1; } }
        @keyframes fadeUp { from { opacity: 0; transform: translateY(30px); } to { opacity: 1; transform: translateY(0); } }
        @keyframes scaleIn { from { opacity: 0; transform: scale(0.8); } to { opacity: 1; transform: scale(1); } }
        @keyframes slideLeft { from { opacity: 0; transform: translateX(50px); } to { opacity: 1; transform: translateX(0); } }
        @keyframes slideRight { from { opacity: 0; transform: translateX(-50px); } to { opacity: 1; transform: translateX(0); } }
        .anim-fadein    { animation: fadeIn    1s   cubic-bezier(0.16, 1, 0.3, 1) forwards; opacity: 0; }
        .anim-fadeup    { animation: fadeUp    1s   cubic-bezier(0.16, 1, 0.3, 1) forwards; opacity: 0; }
        .anim-scale     { animation: scaleIn   0.8s cubic-bezier(0.16, 1, 0.3, 1) forwards; opacity: 0; }
        .anim-slideleft { animation: slideLeft  0.8s cubic-bezier(0.16, 1, 0.3, 1) forwards; opacity: 0; }
        .anim-slideright{ animation: slideRight 0.8s cubic-bezier(0.16, 1, 0.3, 1) forwards; opacity: 0; }`;

// HTML特殊文字をエスケープ
export function escapeHtml(value) {
    return String(value ?? '')
        .replace(/&/g, '&amp;')
        .replace(/</g, '&lt;')
        .replace(/>/g, '&gt;')
        .replace(/"/g, '&quot;')
        .replace(/'/g, '&#039;');
}

// エディタの選択肢と対応する Google Fonts。出力HTMLの<head>へ、使用フォントのみ<link>する。
export const GOOGLE_FONTS = [
    { family: 'Noto Sans JP',      spec: 'Noto+Sans+JP:wght@400;700' },
    { family: 'Noto Serif JP',     spec: 'Noto+Serif+JP:wght@400;700' },
    { family: 'M PLUS Rounded 1c', spec: 'M+PLUS+Rounded+1c:wght@400;700' },
    { family: 'Zen Maru Gothic',   spec: 'Zen+Maru+Gothic:wght@400;700' },
    { family: 'Kosugi Maru',       spec: 'Kosugi+Maru' },
    { family: 'Sawarabi Mincho',   spec: 'Sawarabi+Mincho' },
    { family: 'Yusei Magic',       spec: 'Yusei+Magic' },
    { family: 'Dela Gothic One',   spec: 'Dela+Gothic+One' },
];

export function resolveImageSrc(src, imageMap) {
    if (typeof src === 'string' && src.startsWith('data:image')) {
        return imageMap.get(src) || src;
    }
    return src;
}

// 背景の CSS 宣言を返す。グラデーション on なら background:gradient、それ以外は単色。
export function gradientBgDecl(props, bgcolorEscaped) {
    const g = props.gradient;
    if (g && g.on) {
        const c1 = escapeHtml(g.c1 || '#4facfe'), c2 = escapeHtml(g.c2 || '#00f2fe');
        if (g.type === 'radial') return `background: radial-gradient(circle, ${c1}, ${c2});`;
        const DEG = { v: 180, h: 90, d1: 135, d2: 225 };
        return `background: linear-gradient(${DEG[g.dir] ?? 180}deg, ${c1}, ${c2});`;
    }
    // bgcolor は #rrggbb か rgba(...)（透明度はピッカー内で色に含まれる）
    return `background-color: ${bgcolorEscaped};`;
}

// #rrggbb(または#rgb) と不透明度 → rgba() 文字列
function hexToRgba(hex, a) {
    const h = String(hex ?? '#000000').replace('#', '');
    const n = h.length === 3 ? h.split('').map(c => c + c).join('') : h;
    const r = parseInt(n.slice(0, 2), 16) || 0;
    const g = parseInt(n.slice(2, 4), 16) || 0;
    const b = parseInt(n.slice(4, 6), 16) || 0;
    const al = Math.min(1, Math.max(0, a ?? 1));
    return `rgba(${r}, ${g}, ${b}, ${al})`;
}

// 境界線(Stroke)のCSS宣言。図形/ボタン/画像は border、テキストは -webkit-text-stroke。
export function strokeDecl(props, type) {
    const s = props.stroke;
    if (!s || !s.on) return '';
    const w = Math.max(0, parseFloat(s.width) || 0);
    if (w <= 0) return '';
    const c = escapeHtml(s.color || '#000000');
    if (type === 'Label') return ` -webkit-text-stroke: ${w}px ${c};`;
    return ` border: ${w}px solid ${c};`;
}

// テキストの 斜体/下線/字間/行間 のCSS（font-weight は各要素側で出力）
export function textExtraCss(props) {
    let s = '';
    if (props.italic) s += ' font-style: italic;';
    if (props.underline) s += ' text-decoration: underline;';
    const ls = parseFloat(props.letterSpacing); if (ls) s += ` letter-spacing: ${ls}px;`;
    const lh = parseFloat(props.lineHeight); if (lh) s += ` line-height: ${lh};`;
    return s;
}

// シャドウ系プリセット（種別→CSS値）。Label は text-shadow に使う
const PRESET_SHADOW_CSS = {
    light:    '0 4px 10px rgba(0,0,0,0.15)',
    dark:     '0 8px 15px rgba(0,0,0,0.4)',
    hard:     '5px 5px 0 rgba(0,0,0,0.45)',
    diagonal: '10px 10px 14px rgba(0,0,0,0.3)',
    float:    '0 20px 30px rgba(0,0,0,0.28)',
};

// ドロップシャドウ(自由/プリセット) ＋ 光彩 ＋ 内側シャドウ ＋ ベベル を
// 1つの box-shadow(通常) / text-shadow(Label) に合成して返す。
export function combinedShadowDecl(props, type) {
    const isText = (type === 'Label');
    const box = [];   // box-shadow の各層
    const txt = [];   // text-shadow の各層

    // 1) ドロップシャドウ（自由値優先、無ければプリセット）
    const ds = props.dropShadow;
    if (ds && ds.on) {
        const dx = parseFloat(ds.x) || 0, dy = parseFloat(ds.y) || 0;
        const dblur = Math.max(0, parseFloat(ds.blur) || 0), dspread = parseFloat(ds.spread) || 0;
        const drgba = hexToRgba(ds.color || '#000000', ds.opacity ?? 0.35);
        if (isText) txt.push(`${dx}px ${dy}px ${dblur}px ${drgba}`);              // text-shadow はスプレッド非対応
        else        box.push(`${dx}px ${dy}px ${dblur}px ${dspread}px ${drgba}`);
    } else if (PRESET_SHADOW_CSS[props.shadow]) {
        (isText ? txt : box).push(PRESET_SHADOW_CSS[props.shadow]);
    }

    // 2) 光彩（外側グロー）
    const gl = props.glow;
    if (gl && gl.on) {
        const grgba = hexToRgba(gl.color || '#00d0ff', gl.opacity ?? 0.8);
        const gblur = Math.max(0, parseFloat(gl.blur) || 0), gspread = parseFloat(gl.spread) || 0;
        if (isText) txt.push(`0 0 ${gblur}px ${grgba}`);
        else        box.push(`0 0 ${gblur}px ${gspread}px ${grgba}`);
    }

    // 3) 内側シャドウ（テキスト非対応）
    const is = props.innerShadow;
    if (is && is.on && !isText) {
        const ix = parseFloat(is.x) || 0, iy = parseFloat(is.y) || 0;
        const iblur = Math.max(0, parseFloat(is.blur) || 0);
        const irgba = hexToRgba(is.color || '#000000', is.opacity ?? 0.4);
        box.push(`inset ${ix}px ${iy}px ${iblur}px ${irgba}`);
    }

    // 4) ベベル＆エンボス（テキスト非対応）: 明暗2方向の内側シャドウで立体感
    const bv = props.bevel;
    if (bv && bv.on && !isText) {
        const d = Math.max(1, parseFloat(bv.depth) || 1);
        const op = Math.min(1, Math.max(0, bv.opacity ?? 0.5));
        const hl = hexToRgba(bv.highlight || '#ffffff', op);
        const sh = hexToRgba(bv.shadow || '#000000', op);
        const blur = d * 2;
        if (bv.dir === 'down') {   // 凹（くぼみ）: 左上=影 / 右下=ハイライト
            box.push(`inset ${d}px ${d}px ${blur}px ${sh}`);
            box.push(`inset -${d}px -${d}px ${blur}px ${hl}`);
        } else {                   // 凸（浮き出し）: 左上=ハイライト / 右下=影
            box.push(`inset ${d}px ${d}px ${blur}px ${hl}`);
            box.push(`inset -${d}px -${d}px ${blur}px ${sh}`);
        }
    }

    const arr = isText ? txt : box;
    if (arr.length === 0) return '';
    return (isText ? 'text-shadow: ' : 'box-shadow: ') + arr.join(', ') + ';';
}

// グラデーション文字（文字自体をグラデ塗り）。文字を包む <span> に付けるスタイルを返す。
// off または未設定なら ''。背景クリップと衝突するため必ず span へ適用する。
function gradTextSpanStyle(props) {
    const g = props.gradText;
    if (!g || !g.on) return '';
    const c1 = escapeHtml(g.c1 || '#ff6ec4'), c2 = escapeHtml(g.c2 || '#7873f5');
    const DEG = { v: 180, h: 90, d1: 135, d2: 225 };
    const deg = DEG[g.dir] ?? 90;
    return `background: linear-gradient(${deg}deg, ${c1}, ${c2}); -webkit-background-clip: text; background-clip: text; -webkit-text-fill-color: transparent; color: transparent;`;
}

// テキストをグラデ文字 span で包む（off ならそのまま返す）
export function wrapGradText(text, props) {
    const st = gradTextSpanStyle(props);
    return st ? `<span style="${st}">${text}</span>` : text;
}

File: csharp/SiteBuilder.Core/CssGenerator.cs
// ============================================================
// CssGenerator.cs - HTML/CSS 生成の純粋ヘルパー群
// JS 版 src/renderer/export/css-generator.js の忠実な移植。副作用なし・UI非依存。
// 挙動（既定値・エスケープ・数値フォーマット）を JS と 1:1 で一致させることを目標とする。
// ============================================================
using System.Collections.Generic;
using System.Globalization;

namespace SiteBuilder.Core;

public static class CssGenerator
{
    // アニメーション用CSS（出力HTMLの<style>へ差し込む）。JS: ANIM_CSS
    public const string AnimCss = @"
        @keyframes fadeIn { from { opacity: 0; } to { opacity: 1; } }
        @keyframes fadeUp { from { opacity: 0; transform: translateY(30px); } to { opacity: 1; transform: translateY(0); } }
        @keyframes scaleIn { from { opacity: 0; transform: scale(0.8); } to { opacity: 1; transform: scale(1); } }
        @keyframes slideLeft { from { opacity: 0; transform: translateX(50px); } to { opacity: 1; transform: translateX(0); } }
        @keyframes slideRight { from { opacity: 0; transform: translateX(-50px); } to { opacity: 1; transform: translateX(0); } }
        .anim-fadein    { animation: fadeIn    1s   cubic-bezier(0.16, 1, 0.3, 1) forwards; opacity: 0; }
        .anim-fadeup    { animation: fadeUp    1s   cubic-bezier(0.16, 1, 0.3, 1) forwards; opacity: 0; }
        .anim-scale     { animation: scaleIn   0.8s cubic-bezier(0.16, 1, 0.3, 1) forwards; opacity: 0; }
        .anim-slideleft { animation: slideLeft  0.8s cubic-bezier(0.16, 1, 0.3, 1) forwards; opacity: 0; }
        .anim-slideright{ animation: slideRight 0.8s cubic-bezier(0.16, 1, 0.3, 1) forwards; opacity: 0; }";

    // エディタの選択肢と対応する Google Fonts。JS: GOOGLE_FONTS
    public static readonly IReadOnlyList<GoogleFont> GoogleFonts = new[]
    {
        new GoogleFont("Noto Sans JP",      "Noto+Sans+JP:wght@400;700"),
        new GoogleFont("Noto Serif JP",     "Noto+Serif+JP:wght@400;700"),
        new GoogleFont("M PLUS Rounded 1c", "M+PLUS+Rounded+1c:wght@400;700"),
        new GoogleFont("Zen Maru Gothic",   "Zen+Maru+Gothic:wght@400;700"),
        new GoogleFont("Kosugi Maru",       "Kosugi+Maru"),
        new GoogleFont("Sawarabi Mincho",   "Sawarabi+Mincho"),
        new GoogleFont("Yusei Magic",       "Yusei+Magic"),
        new GoogleFont("Dela Gothic One",   "Dela+Gothic+One"),
    };

    // HTML特殊文字をエスケープ。JS: escapeHtml
    public static string EscapeHtml(string? value)
    {
        return (value ?? string.Empty)
            .Replace("&", "&amp;")
            .Replace("<", "&lt;")
            .Replace(">", "&gt;")
            .Replace("\"", "&quot;")
            .Replace("'", "&#039;");
    }

    // data:image を imageMap で解決。JS: resolveImageSrc
    public static string? ResolveImageSrc(string? src, IReadOnlyDictionary<string, string> imageMap)
    {
        if (src is not null && src.StartsWith("data:image", System.StringComparison.Ordinal))
            return imageMap.TryGetValue(src, out var mapped) ? mapped : src;
        return src;
    }

    // 背景の CSS 宣言。グラデーション on なら background:gradient、それ以外は単色。JS: gradientBgDecl
    public static string GradientBgDecl(ElementProps props, string bgcolorEscaped)
    {
        var g = props.Gradient;
        if (g is { On: true })
        {
            var c1 = EscapeHtml(g.C1 ?? "#4facfe");
            var c2 = EscapeHtml(g.C2 ?? "#00f2fe");
            if (g.Type == "radial")
                return $"background: radial-gradient(circle, {c1}, {c2});";
            return $"background: linear-gradient({DirDeg(g.Dir, 180)}deg, {c1}, {c2});";
        }
        // bgcolor は #rrggbb か rgba(...)（透明度はピッカー内で色に含まれる）
        return $"background-color: {bgcolorEscaped};";
    }

    // 境界線(Stroke)のCSS宣言。図形/ボタン/画像は border、テキストは -webkit-text-stroke。JS: strokeDecl
    public static string StrokeDecl(ElementProps props, string? type)
    {
        var s = props.Stroke;
        if (s is not { On: true }) return string.Empty;
        var w = System.Math.Max(0, s.Width ?? 0);
        if (w <= 0) return string.Empty;
        var c = EscapeHtml(s.Color ?? "#000000");
        if (type == "Label") return $" -webkit-text-stroke: {Num(w)}px {c};";
        return $" border: {Num(w)}px solid {c};";
    }

    // テキストの 斜体/下線/字間/行間 のCSS（font-weight は各要素側で出力）。JS: textExtraCss
    public static string TextExtraCss(ElementProps props)
    {
        var s = string.Empty;
        if (props.Italic) s += " font-style: italic;";
        if (props.Underline) s += " text-decoration: underline;";
        var ls = props.LetterSpacing ?? 0; if (ls != 0) s += $" letter-spacing: {Num(ls)}px;";
        var lh = props.LineHeight ?? 0; if (lh != 0) s += $" line-height: {Num(lh)};";
        return s;
    }

    // シャドウ系プリセット（種別→CSS値）。Label は text-shadow に使う。JS: PRESET_SHADOW_CSS
    private static readonly IReadOnlyDictionary<string, string> PresetShadowCss = new Dictionary<string, string>
    {
        ["light"]    = "0 4px 10px rgba(0,0,0,0.15)",
        ["dark"]     = "0 8px 15px rgba(0,0,0,0.4)",
        ["hard"]     = "5px 5px 0 rgba(0,0,0,0.45)",
        ["diagonal"] = "10px 10px 14px rgba(0,0,0,0.3)",
        ["float"]    = "0 20px 30px rgba(0,0,0,0.28)",
    };

    // ドロップシャドウ ＋ 光彩 ＋ 内側シャドウ ＋ ベベル を
    // 1つの box-shadow(通常) / text-shadow(Label) に合成。JS: combinedShadowDecl
    public static string CombinedShadowDecl(ElementProps props, string? type)
    {
        var isText = type == "Label";
        var box = new List<string>();
        var txt = new List<string>();

        // 1) ドロップシャドウ（自由値優先、無ければプリセット）
        var ds = props.DropShadow;
        if (ds is { On: true })
        {
            var dx = ds.X ?? 0; var dy = ds.Y ?? 0;
            var dblur = System.Math.Max(0, ds.Blur ?? 0); var dspread = ds.Spread ?? 0;
            var drgba = HexToRgba(ds.Color ?? "#000000", ds.Opacity ?? 0.35);
            if (isText) txt.Add($"{Num(dx)}px {Num(dy)}px {Num(dblur)}px {drgba}");           // text-shadow はスプレッド非対応
            else        box.Add($"{Num(dx)}px {Num(dy)}px {Num(dblur)}px {Num(dspread)}px {drgba}");
        }
        else if (props.Shadow is not null && PresetShadowCss.TryGetValue(props.Shadow, out var preset))
        {
            (isText ? txt : box).Add(preset);
        }

        // 2) 光彩（外側グロー）
        var gl = props.Glow;
        if (gl is { On: true })
        {
            var grgba = HexToRgba(gl.Color ?? "#00d0ff", gl.Opacity ?? 0.8);
            var gblur = System.Math.Max(0, gl.Blur ?? 0); var gspread = gl.Spread ?? 0;
            if (isText) txt.Add($"0 0 {Num(gblur)}px {grgba}");
            else        box.Add($"0 0 {Num(gblur)}px {Num(gspread)}px {grgba}");
        }

        // 3) 内側シャドウ（テキスト非対応）
        var iss = props.InnerShadow;
        if (iss is { On: true } && !isText)
        {
            var ix = iss.X ?? 0; var iy = iss.Y ?? 0;
            var iblur = System.Math.Max(0, iss.Blur ?? 0);
            var irgba = HexToRgba(iss.Color ?? "#000000", iss.Opacity ?? 0.4);
            box.Add($"inset {Num(ix)}px {Num(iy)}px {Num(iblur)}px {irgba}");
        }

        // 4) ベベル＆エンボス（テキスト非対応）: 明暗2方向の内側シャドウで立体感
        var bv = props.Bevel;
        if (bv is { On: true } && !isText)
        {
            var d = System.Math.Max(1, bv.Depth ?? 1);
            var op = System.Math.Min(1, System.Math.Max(0, bv.Opacity ?? 0.5));
            var hl = HexToRgba(bv.Highlight ?? "#ffffff", op);
            var sh = HexToRgba(bv.Shadow ?? "#000000", op);
            var blur = d * 2;
            if (bv.Dir == "down")   // 凹（くぼみ）: 左上=影 / 右下=ハイライト
            {
                box.Add($"inset {Num(d)}px {Num(d)}px {Num(blur)}px {sh}");
                box.Add($"inset -{Num(d)}px -{Num(d)}px {Num(blur)}px {hl}");
            }
            else                    // 凸（浮き出し）: 左上=ハイライト / 右下=影
            {
                box.Add($"inset {Num(d)}px {Num(d)}px {Num(blur)}px {hl}");
                box.Add($"inset -{Num(d)}px -{Num(d)}px {Num(blur)}px {sh}");
            }
        }

        var arr = isText ? txt : box;
        if (arr.Count == 0) return string.Empty;
        return (isText ? "text-shadow: " : "box-shadow: ") + string.Join(", ", arr) + ";";
    }

    // テキストをグラデ文字 span で包む（off ならそのまま返す）。JS: wrapGradText
    public static string WrapGradText(string text, ElementProps props)
    {
        var st = GradTextSpanStyle(props);
        return st.Length > 0 ? $"<span style=\"{st}\">{text}</span>" : text;
    }

    // グラデーション文字（文字自体をグラデ塗り）。span に付けるスタイルを返す。JS: gradTextSpanStyle
    private static string GradTextSpanStyle(ElementProps props)
    {
        var g = props.GradText;
        if (g is not { On: true }) return string.Empty;
        var c1 = EscapeHtml(g.C1 ?? "#ff6ec4");
        var c2 = EscapeHtml(g.C2 ?? "#7873f5");
        var deg = DirDeg(g.Dir, 90);
        return $"background: linear-gradient({deg}deg, {c1}, {c2}); -webkit-background-clip: text; background-clip: text; -webkit-text-fill-color: transparent; color: transparent;";
    }

    // #rrggbb(または#rgb) と不透明度 → rgba() 文字列。JS: hexToRgba
    private static string HexToRgba(string? hex, double a)
    {
        var h = (hex ?? "#000000").Replace("#", string.Empty);
        var n = h.Length == 3
            ? new string(new[] { h[0], h[0], h[1], h[1], h[2], h[2] })
            : h;
        var r = HexByte(n, 0);
        var g = HexByte(n, 2);
        var b = HexByte(n, 4);
        var al = System.Math.Min(1, System.Math.Max(0, a));
        return $"rgba({r}, {g}, {b}, {Num(al)})";
    }

    // n[start..start+2] を16進で読む。範囲外・不正は 0（JS の parseInt(...,16) || 0 と等価）
    private static int HexByte(string n, int start)
    {
        if (start >= n.Length) return 0;
        var len = System.Math.Min(2, n.Length - start);
        var slice = n.Substring(start, len);
        return int.TryParse(slice, NumberStyles.HexNumber, CultureInfo.InvariantCulture, out var v) ? v : 0;
    }

    // グラデ方向キー → 角度（度）。JS: DEG = { v:180, h:90, d1:135, d2:225 } と fallback
    private static int DirDeg(string? dir, int fallback) => dir switch
    {
        "v" => 180,
        "h" => 90,
        "d1" => 135,
        "d2" => 225,
        _ => fallback,
    };

    // JS のテンプレートリテラル ${number} 相当の数値フォーマット（ロケール非依存、末尾ゼロなし）
    private static string Num(double v) => v.ToString("0.############", CultureInfo.InvariantCulture);
}
